clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/subscriber/CachingPubSubSubscriberFactory.java

```diff
@@ -0,0 +1,116 @@
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import org.eclipse.hono.client.ServerErrorException;
+import org.eclipse.hono.client.pubsub.PubSubMessageHelper;
+
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.cloud.pubsub.v1.MessageReceiver;
+
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+
+/**
+ * A factory for creating PubSubSubscribers. Created subscribers are being cached.
+ */
+public class CachingPubSubSubscriberFactory implements PubSubSubscriberFactory {
+
+    private final Vertx vertx;
+    private final Map<String, PubSubSubscriber> activeSubscribers = new ConcurrentHashMap<>();
+    private final String projectId;
+    private final FixedCredentialsProvider credentialsProvider;
+    private Supplier<PubSubSubscriber> clientSupplier;
+
+    /**
+     * Creates a new factory for {@link PubSubSubscriber} instances.
+     *
+     * @param vertx The Vert.x instance that this factory runs on.
+     * @param projectId The identifier of the Google Cloud Project to connect to.
+     * @param credentialsProvider The provider for credentials to use for authenticating to the Pub/Sub service.
+     * @throws NullPointerException If any of the parameters is {@code null}.
+     */
+    public CachingPubSubSubscriberFactory(
+            final Vertx vertx,
+            final String projectId,
+            final FixedCredentialsProvider credentialsProvider) {
+        this.vertx = Objects.requireNonNull(vertx);
+        this.projectId = Objects.requireNonNull(projectId);
+        this.credentialsProvider = Objects.requireNonNull(credentialsProvider);
+    }
+
+    /**
+     * Sets a supplier for the subscriber(s) this factory creates.
+     * <p>
+     * This method is mainly intended to be used in test cases.
+     *
+     * @param supplier The supplier.
+     */
+    public void setClientSupplier(final Supplier<PubSubSubscriber> supplier) {
+        this.clientSupplier = supplier;
+    }
+
+    @Override
+    public Future<Void> closeSubscriber(final String subscription, final String prefix) {
+        final String subscriptionId = PubSubMessageHelper.getTopicName(subscription, prefix);
+        return removeSubscriber(subscriptionId);
+    }
+
+    @Override
+    public Future<Void> closeAllSubscriber() {
+        activeSubscribers.forEach((k, v) -> removeSubscriber(k));
+        if (activeSubscribers.isEmpty()) {
+            return Future.succeededFuture();
+        }
+        return Future.failedFuture(
+                new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "Failed to close all subscriber"));
+    }
+
+    @Override
+    public PubSubSubscriber getOrCreateSubscriber(final String subscriptionId, final MessageReceiver receiver) {
+        return activeSubscribers.computeIfAbsent(subscriptionId,
+                s -> createPubSubSubscriber(subscriptionId, receiver));
```

Comment:
Yes, the existing subscriber will be returned if it already exists, because there should be only one subscriber for our subscription. Also later in the other PRs, you will see that it is always the same message receiver, because this function is only called with the same receiver and subscription Id by the same caller. There is no case that another caller calls this function with another MessageReceiver and the same topic.